### PR TITLE
DEMO 326 (Feat): Add state to input panel

### PR DESCRIPTION
**Pull request by FMorrison87**
**JIRA Ticket [Jira Issue Key](ticketURL)**

## Brief Description

This PR pulls in state data and populates the form in the input data panel. The database file input when selected displays all possible trackfiles for the selected spacecraft. On selection, the trackfileId is sent back via store dispatch and then all trakfile values in the form should update according to trackfile state. `generateprocessedTrackFile` is also pulled in from the mock data via a service and is used to generate a processed track file per track file when Determine Orbit is clicked. I also added a new trackfile action in the reducer and under actions to add in a processed trackfile object to a selected trackfile.

This PR includes a new custom selector to get the all track files on a spacecraft as an array of trackfile objects.

## Issues and/or Limitations

I'm not 100% sure that the values being passed into the form are correct as the names don't match 1:1 with what's in the wireframe.

I did not engineer the type-ahead design wanted for the intial database file dropdown, as I wasn't sure if it was in scope for this ticket.

## Final Checklist

- [ ] Works in Chrome
- [ ] Works in Firefox
- [ ] Works in Safari
- [ ] Handles responsiveness as required
- [ ] Dark theme styles are correct
- [ ] Light theme styles are correct

### CODE REVIEW

File: src/app/main/inputs-outputs/inputs/inputs.component.ts [hunk 1]
```diff
@@ -11,23 +21,70 @@ import { FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
...
+  }>;
+  processedTrackFile: ProcessedTrackFile | undefined;
+  currentTrackFileId: string | undefined;
+  trackfiles$: Observable<TrackFile[] | undefined> = this.store.select(
+    selectCurrentSpaceCraftTrackFiles
+  );
+  currentTrackFile$: Subscription = this.store
+    .pipe(
+      select(selectCurrentTrackFile),
+      filter((val) => val !== null)
```
Comment (NicDominguez) [inline src/app/main/inputs-outputs/inputs/inputs.component.ts:41]: This was the pattern I was looking for. I didn't realize you could use select inside a pipe.

**Review by Paterma — APPROVED** [review]

File: src/app/main/inputs-outputs/inputs/inputs.component.ts [hunk 2]
```diff
@@ -11,23 +21,70 @@ import { FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
...
-    console.log(this.inputForm.value);
-    //TODO hook form data into where it's going to go
+    // generate processed trackFile using service
+    (this.processedTrackFile = this.ProcessTrackFileService.processtrackFile(
+      this.currentTrackFileId!
+    )),
+      //Dispatch processed trackfile as a property of the current trackfile, back into state
+      this.store.dispatch(
+        TrackFilesActions.trackFileProcessed({
+          trackFileId: this.currentTrackFileId!,
```
Comment (NicDominguez) [inline src/app/main/inputs-outputs/inputs/inputs.component.ts:77]: This trackFileId is not necessary sense processedTrackFile has refId attached

File: src/app/core/scenario-library/scenario-library.component.ts [hunk 3]
```diff
@@ -22,47 +26,51 @@ import { Observable, Subscription } from 'rxjs';
...
-    });
-
-    this.spacecraftSub = this.selectedSpacecraftId$.subscribe((res: string | null) => {
-      this.selectedSpacecraftId = res;
-    });
+    this.spacecraftSub = this.selectedSpacecraftId$.subscribe(
+      (res: string | null) => {
+        this.selectedSpacecraftId = res;
+      }
+    );
```
Comment (NicDominguez) [inline src/app/core/scenario-library/scenario-library.component.ts:64]: Please change to new pattern

File: src/app/core/scenario-library/scenario-library.component.ts [hunk 4]
```diff
@@ -22,47 +26,51 @@ import { Observable, Subscription } from 'rxjs';
...
 
     this.spacecraftsSub = this.spacecrafts$.subscribe((res: any) => {
       this.spacecraftData = res;
     });
   }
```
Comment (NicDominguez) [inline src/app/core/scenario-library/scenario-library.component.ts:69]: Please change to new pattern

File: src/app/core/scenario-library/scenario-library.component.ts [hunk 5]
```diff
@@ -22,47 +26,51 @@ import { Observable, Subscription } from 'rxjs';
...
   scenarios: (Scenario | undefined)[] = [];
+  scenarios$: Subscription = this.store
+    .select(selectScenarios)
+    .pipe(
+      map(
+        (scenario) =>
+          (this.scenarios = scenario.ids.map((id) => {
+            return scenario.entities[id];
+          }))
+      )
```
Comment (NicDominguez) [inline src/app/core/scenario-library/scenario-library.component.ts:38]: Check with Jacob to see if using Object.values(scenarioEntity) is doable

File: src/app/+state/app.selectors.ts [hunk 6]
```diff
@@ -57,3 +57,36 @@ export const selectCurrentTrackFile = createSelector(
...
+    const spacecraftTrackfileIds: string[] = spacecraft.trackFileIds;
+    let spacecraftTrackfileNames = [];
+    if (spacecraftTrackfileIds) {
+      for (const id of spacecraftTrackfileIds) {
+        spacecraftTrackfileNames.push(trackfiles[id]?.name);
+      }
+    }
+    return spacecraftTrackfileNames;
+  }
+);
```
Comment (NicDominguez) [inline src/app/+state/app.selectors.ts:75]: Possibly remove

File: src/app/+state/app.selectors.ts [hunk 7]
```diff
@@ -57,3 +57,36 @@ export const selectCurrentTrackFile = createSelector(
...
+    const spacecraftTrackfileIds: string[] = spacecraft.trackFileIds;
+    let spacecraftAllTrackFiles: TrackFile[] = [];
+    if (spacecraftTrackfileIds) {
+      for (const id of spacecraftTrackfileIds) {
+        spacecraftAllTrackFiles.push(trackfiles[id]!);
+      }
+    }
+    return spacecraftAllTrackFiles;
+  }
+);
```
Comment (NicDominguez) [inline src/app/+state/app.selectors.ts:92]: Possibly remove

**Review by NicDominguez — CHANGES_REQUESTED** [review]

File: src/app/core/scenario-library/scenario-library.component.ts (hunk 3 above)
Comment (FMorrison87) [reply src/app/core/scenario-library/scenario-library.component.ts]: Most of this has been removed now.

File: src/app/core/scenario-library/scenario-library.component.ts (hunk 4 above)
Comment (FMorrison87) [reply src/app/core/scenario-library/scenario-library.component.ts]: This was removed.

File: src/app/main/inputs-outputs/inputs/inputs.component.ts (hunk 2 above)
Comment (FMorrison87) [reply src/app/main/inputs-outputs/inputs/inputs.component.ts:77]: all set

File: src/app/mock-data/generate-data.ts [hunk 8]
```diff
@@ -85,27 +85,26 @@ const generateEphemerisFile = (
...
     line2:
       '2 25544  51.6416 224.4564 0005880  33.3950 353.6634 15.49394965416327',
   };
 };
 
-const generateProcessedTrackFile = (
-  trackFileId: string,
-  index: number
+export const generateProcessedTrackFile = (
+  trackFileId: string | null
```
Comment (FMorrison87) [inline src/app/mock-data/generate-data.ts:97]: I fixed this too, took out the null

File: src/app/types/data.types.ts [hunk 9]
```diff
@@ -61,7 +61,7 @@ export type TrackFileEntity = {
 };
 
 export type ProcessedTrackFile = {
-  trackFileRefId: string;
+  trackFileRefId: string | null;
```
Comment (FMorrison87) [inline src/app/types/data.types.ts:64]: I removed the null

**Review by NicDominguez — APPROVED** [review]

**Pull request merged**